### PR TITLE
Fix mapping records in nested namespaces (#8681)

### PR DESCRIPTION
**Pull request by MpDzik**
PR for #8681

### CODE REVIEW

File: src/Orchard/Environment/ShellBuilders/CompositionStrategy.cs [hunk 1]
```diff
@@ -197,7 +197,7 @@ public class CompositionStrategy : ICompositionStrategy {
         }
 
         private static bool IsRecord(Type type) {
-            return ((type.Namespace ?? "").EndsWith(".Models") || (type.Namespace ?? "").EndsWith(".Records")) &&
+            return ((type.Namespace ?? "").EndsWith(".Models") || (type.Namespace ?? "").EndsWith(".Records") || type.Name.EndsWith("Record")) &&
```
Comment (BenedekFarkas) [inline src/Orchard/Environment/ShellBuilders/CompositionStrategy.cs]: Would it matter if we detect a class to be a record even when it isn't (i.e., false positive)?
In any case, we can just modify this to be more forgiving like this:
```suggestion
            var namespace = type.Namespace ?? string.Empty;
            return (namespace.Contains(".Models") || namespace.Contains(".Records")) &&
```

Comment (MatteoPiovanelli) [reply src/Orchard/Environment/ShellBuilders/CompositionStrategy.cs]: The chance for false positive makes this PR potentially breaking. AFAIK it wouldn't break anything in the current Orchard codebase, nor ours, but the potential is there.

The approach using `Contains` may be more robust. The proposed snippet is missing a few corner cases (e.g. the "Example.ModelsX" namespace would pass the test now, while it never did and probably shouldn't). A possible middle ground could be to have both tests?
```
// Only for "Models" namespace and its sub-namespaces, but similarly for Records:
(namespace.Contains(".Models.") || namespace.EndsWith(".Models"))
```

Comment (MpDzik) [reply src/Orchard/Environment/ShellBuilders/CompositionStrategy.cs]: It actually does cause issues if a non-record class is mapped. For example, if the class contains non-virtual methods or doesn't have a parameterless constructor and Orchard tries to map it with NHibernate, it will throw an exception. This already happens if you put a class which is not a valid record into the `Models` folder and add a virtual `Id` property.

I like @MatteoPiovanelli-Laser's approach with checking the exact name `Models` or `Records`. It is consistent with the current naming convention. If we want to by fully backward compatible, we could add an attribute to opt-in which classes should be mapped. The drawback of this approach is that it would be inconsistent with the records in non-nested namespaces, so we would need the attribute only for the nested namespaces. And frankly, why would anyone name a non-record class with a `Record` suffix?

Maybe it would make sense to use the `DoNotMap` attribute to exclude classes from mapping. This way, even if this change breaks any custom modules, it will be easy to resolve the issue by adding the attribute to the class that was mapped as a false positive.

Comment (MatteoPiovanelli) [reply src/Orchard/Environment/ShellBuilders/CompositionStrategy.cs]: The opt-in attribute makes sense for classes outside the current naming convention, I think. 
Perhaps something like `[MapAsRecord]` for the classes that should be tentatively be considered records; when such class is in a Namespace that matches the conventions, such attribute should be redundant, for backwards compatibility.
At the same time `[MapAsRecord(false)]` could prevent classes that shouldn't be considered records to be considered as such even if they happened to satisfy all other constraints.

Comment (MpDzik) [reply src/Orchard/Environment/ShellBuilders/CompositionStrategy.cs]: Yes, I was thinking about the same. It seems strange to me that it was not implemented this way from the beginning. Using an attribute or marker interface is a lot better than relying on the namespace layout, in my opinion.

**Comment by sebastienros** [comment]
I'll let you all agree on the best solution (which doesn't break anything) since you are more knowledgeable.

**Review by sebastienros — APPROVED** [review]

**Comment by MatteoPiovanelli** [comment]
@MpDzik could you update the PR with the things we discussed above?

**Comment by MpDzik** [comment]
@MatteoPiovanelli-Laser How does this look to you?

File: src/Orchard/Environment/ShellBuilders/CompositionStrategy.cs (hunk 1 above)
Comment (BenedekFarkas) [reply src/Orchard/Environment/ShellBuilders/CompositionStrategy.cs]: Looks great, thanks!

**Review by BenedekFarkas — APPROVED** [review]

**Review by MatteoPiovanelli — APPROVED** [review]

**Comment by MatteoPiovanelli** [comment]
sorry for the delay on my part, but I had missed the notification.
